### DISCUSSION TITLE
Plugin architecture

### DISCUSSION THREAD

**Discussion by billz** [RaspAP/raspap-webgui#1052] [Ideas]
One idea I've played with is adding plugin support for RaspAP. This would let other developers build enhancements for the project without modifying the core code.

Examples include this [interesting fork](https://github.com/Erik1000/raspap-webgui/commit/ebb73e41d1929983a312367c93bc86ea139c3f78) of RaspAP that adds [RPiPlay](https://github.com/FD-/RPiPlay) integration. While this may not appeal to everyone, there are probably some users who would add this to RaspAP if it could be done simply.

There are several design patterns for a plugin system that could be used here. My personal take is to avoid frameworks and go with a minimal event dispatcher-type implementation.

Does a plugin system make sense for RaspAP? 
If yes, which plugins would you like to see added?

**Comment by assachs** [comment]
Hello Billz,
i want to push this idea. Some rudimentary plugin support would be nice. This includes 

- adding entries to the sidebar
- delegating page_actions to the plugin
- render templates from a plugin directory

I already implemented a first version of the plugin support.

My personal use case:
we use raspap on a DIY electronic scoreboard to connect the scoreboard to the wifi and also to configure the scoreboard (this is the plugin part)

![image](https://github.com/user-attachments/assets/8d518faf-b822-4c4a-a3da-4bdcc2f8c8ef)

![image](https://github.com/user-attachments/assets/9392499f-03d2-4f20-9c4f-ebe2599cf859)

At first i changed raspap and included the changes for the scoreboard. For the second version i added plugin support to raspap and moved the scoreboard functionality to the plugin. But it would be nice, if the plugin support is available in the original version and not only in my fork.

**Reply by billz** [reply]
Glad to see this and nicely done. I'll dust off my plugin fork and perhaps integrate some of the work you've already done.

**Comment by billz** [comment]
Currently in progress #1693. User plugins are auto-loaded from the `/plugins` folder by the PluginManager.

A priority value allows plugin sidebar items to be positioned arbitrarily:

<img width="221" alt="image" src="https://github.com/user-attachments/assets/80a1a2ab-7d5d-4089-ae0a-eb3ca921bf58">

I'll publish SamplePlugin in a separate repo shortly, once the template rendering and page action handling is done.

**Comment by assachs** [comment]
Your solution looks very nice. Thank you very much.

**Reply by billz** [reply]
@assachs your work provided some much needed inspiration. I will credit you in the code and release notes :)

**Comment by assachs** [comment]
@billz Hello Bill, i had no time to convert my plugin to the new plugin implementation (maybe tomorrow), but i already looked at the source code.

2 Ideas:
- Move handlePageAction to the Plugin-Class. 
- The class autoloader should also use the pluginmanager and should load classes with the same namespace (or subdirectory) from the plugin-folder.

With this approach functions.php is not needed any more. One Namespace is sufficient (at the moment there is a namespace for functions.php and one for the Plugin).

For complex plugins the functions can be divided in several classes without the need of require_once

**Reply by billz** [reply]
@assachs thanks for the valuable feedback. I agree re: handlePageAction. Actually, I started coding this in the PluginManager and later implemented it again in the [SamplePlugin](https://github.com/RaspAP/SamplePlugin), so arrived at this in a roundabout way. Encapsulating this at the plugin level is a better approach.

**Comment by billz** [comment]
Both the PluginManager and [SamplePlugin](https://github.com/RaspAP/SamplePlugin) are nearly code complete. The latter needs some usage notes in the README, although the code is reasonably well documented.

<img width="404" alt="sample" src="https://github.com/user-attachments/assets/06c14fbd-2510-41ed-a523-7a5575ee0e08">

<img width="619" alt="multiple" src="https://github.com/user-attachments/assets/26110039-b9dc-4ca3-837e-739f25ace3c9">

The PluginManager was tested with multiple plugin instances, each with their own data persistence.
